File: java/app/src/main/java/com/jossy/android/mobilebenchmarkappjava/metrics/SystemMetricsCollector.java
package com.jossy.android.mobilebenchmarkappjava.metrics;

import android.app.ActivityManager;
import android.content.Context;
import android.os.Debug;
import android.os.Handler;
import android.os.Looper;
import android.os.Process;
import android.util.Log;
import android.view.Choreographer;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Kolektor metryk systemowych - zbiera CPU, RAM, GPU, FPS bezpośrednio z
 * aplikacji.
 * Zapisuje dane do CSV w czasie rzeczywistym.
 */
public class SystemMetricsCollector {

    private static final String TAG = "SystemMetricsCollector";

    // Konfiguracja
    private final int samplingIntervalMs;
    private static final int BUFFER_SIZE = 50; // Flush co 50 próbek

    private final Context context;
    private final File outputFile;
    private final ExecutorService executor;
    private final Handler mainHandler;

    // Stan
    private final AtomicBoolean running = new AtomicBoolean(false);
    private String currentTestName = "IDLE";
    private int currentIteration = 0;
    private long startTimeMs = 0;

    // Aktualne wartości dla UI
    private volatile float lastCpuUsage = 0f;
    private volatile float lastMemoryUsage = 0f;
    private volatile float currentFps = 0f;

    // Metryki CPU
    private long prevCpuTime = 0;
    private long prevAppCpuTime = 0;

    // Metryki FPS
    private final AtomicInteger frameCount = new AtomicInteger(0);
    private final AtomicLong lastFpsCalcTime = new AtomicLong(0);

    // Bufor i writer
    private BufferedWriter writer;
    private final List<MetricSample> buffer = new ArrayList<>(BUFFER_SIZE);

    // Choreographer callback dla FPS
    private Choreographer.FrameCallback frameCallback;

    public SystemMetricsCollector(Context context, File outputDir, String sessionId, int samplingIntervalMs) {
        this.context = context.getApplicationContext();
        this.outputFile = new File(outputDir, "system_metrics_" + sessionId + ".csv");
        this.samplingIntervalMs = samplingIntervalMs;
        this.executor = Executors.newSingleThreadExecutor(r -> {
            Thread t = new Thread(r, "metrics-collector");
            t.setPriority(Thread.MIN_PRIORITY);
            return t;
        });
        this.mainHandler = new Handler(Looper.getMainLooper());
    }

    public void start() throws IOException {
        if (running.getAndSet(true)) {
            return;
        }

        if (!outputFile.getParentFile().exists()) {
            outputFile.getParentFile().mkdirs();
        }
        writer = new BufferedWriter(new FileWriter(outputFile), 32 * 1024);
        writer.write("TimestampMs,ElapsedMs,TestName,Iteration,CPU%,RAM_MB,GPU%,FPS,PID\n");

        startTimeMs = System.currentTimeMillis();
        lastFpsCalcTime.set(startTimeMs);

        startFpsCounter();
        executor.submit(this::samplingLoop);

        Log.i(TAG, "Started metrics collection: " + outputFile.getAbsolutePath());
    }

    public void stop() {
        if (!running.getAndSet(false)) {
            return;
        }

        stopFpsCounter();
        executor.shutdown();
        try {
            executor.awaitTermination(5, TimeUnit.SECONDS);
        } catch (InterruptedException ignored) {
        }

        try {
            synchronized (buffer) {
                flushBuffer();
            }
            if (writer != null) {
                writer.close();
            }
        } catch (IOException e) {
            Log.e(TAG, "Error closing writer", e);
        }

        Log.i(TAG, "Stopped metrics collection");
    }

    public void setCurrentTest(String testName, int iteration) {
        this.currentTestName = testName;
        this.currentIteration = iteration;
    }

    public float getLastCpuUsage() {
        return lastCpuUsage;
    }

    public float getLastMemoryUsage() {
        return lastMemoryUsage;
    }

    public float getLastFps() {
        return currentFps;
    }

    private void samplingLoop() {
        while (running.get()) {
            long loopStart = System.currentTimeMillis();

            try {
                MetricSample sample = collectSample();
                // Aktualizuj wartości dla UI
                lastCpuUsage = sample.cpuPercent;
                lastMemoryUsage = sample.ramMb;

                synchronized (buffer) {
                    buffer.add(sample);
                    if (buffer.size() >= BUFFER_SIZE) {
                        flushBuffer();
                    }
                }
            } catch (Exception e) {
                Log.e(TAG, "Sampling error", e);
            }

            long elapsed = System.currentTimeMillis() - loopStart;
            long sleepMs = Math.max(1, samplingIntervalMs - elapsed);
            try {
                Thread.sleep(sleepMs);
            } catch (InterruptedException ignored) {
                break;
            }
        }
    }

    private MetricSample collectSample() {
        long now = System.currentTimeMillis();
        return new MetricSample(
                now,
                now - startTimeMs,
                currentTestName,
                currentIteration,
                getCpuUsage(),
                getMemoryUsageMB(),
                getGpuUsage(),
                currentFps,
                Process.myPid());
    }

    private float getCpuUsage() {
        try {
            int pid = Process.myPid();
            RandomAccessFile procStat = new RandomAccessFile("/proc/" + pid + "/stat", "r");
            String statLine = procStat.readLine();
            procStat.close();

            String[] parts = statLine.split("\\s+");
            if (parts.length < 15) return 0f;

            long utime = Long.parseLong(parts[13]);
            long stime = Long.parseLong(parts[14]);
            long appCpuTime = utime + stime;

            RandomAccessFile cpuStat = new RandomAccessFile("/proc/stat", "r");
            String cpuLine = cpuStat.readLine();
            cpuStat.close();

            String[] cpuParts = cpuLine.split("\\s+");
            long cpuTime = 0;
            for (int i = 1; i <= 7 && i < cpuParts.length; i++) {
                cpuTime += Long.parseLong(cpuParts[i]);
            }

            if (prevCpuTime > 0 && prevAppCpuTime > 0) {
                long cpuDelta = cpuTime - prevCpuTime;
                long appDelta = appCpuTime - prevAppCpuTime;
                if (cpuDelta > 0) {
                    int cores = Runtime.getRuntime().availableProcessors();
                    float percent = (float) appDelta / cpuDelta * 100f * cores;
                    prevCpuTime = cpuTime;
                    prevAppCpuTime = appCpuTime;
                    return Math.min(100f * cores, percent);
                }
            }

            prevCpuTime = cpuTime;
            prevAppCpuTime = appCpuTime;
            return 0f;

        } catch (Exception e) {
            return 0f;
        }
    }

    private float getMemoryUsageMB() {
        try {
            Debug.MemoryInfo memInfo = new Debug.MemoryInfo();
            Debug.getMemoryInfo(memInfo);
            return memInfo.getTotalPss() / 1024f;
        } catch (Exception e) {
            return 0f;
        }
    }

    private float getGpuUsage() {
        try {
            BufferedReader reader = new BufferedReader(new FileReader("/sys/class/kgsl/kgsl-3d0/gpu_busy_percentage"));
            String line = reader.readLine();
            reader.close();
            if (line != null) return Float.parseFloat(line.trim());
        } catch (Exception ignored) {}
        return -1f;
    }

    private void startFpsCounter() {
        frameCallback = new Choreographer.FrameCallback() {
            @Override
            public void doFrame(long frameTimeNanos) {
                if (!running.get()) return;
                frameCount.incrementAndGet();
                long now = System.currentTimeMillis();
                long lastCalc = lastFpsCalcTime.get();
                if (now - lastCalc >= 1000) {
                    int frames = frameCount.getAndSet(0);
                    float interval = (now - lastCalc) / 1000f;
                    currentFps = frames / interval;
                    lastFpsCalcTime.set(now);
                }
                Choreographer.getInstance().postFrameCallback(this);
            }
        };
        mainHandler.post(() -> Choreographer.getInstance().postFrameCallback(frameCallback));
    }

    private void stopFpsCounter() {
        mainHandler.post(() -> {
            if (frameCallback != null) Choreographer.getInstance().removeFrameCallback(frameCallback);
        });
    }

    private void flushBuffer() throws IOException {
        if (buffer.isEmpty() || writer == null) return;
        StringBuilder sb = new StringBuilder();
        for (MetricSample s : buffer) {
            sb.append(String.format(Locale.US, "%d,%d,%s,%d,%.2f,%.2f,%.2f,%.1f,%d\n",
                    s.timestampMs, s.elapsedMs, s.testName, s.iteration,
                    s.cpuPercent, s.ramMb, s.gpuPercent, s.fps, s.pid));
        }
        writer.write(sb.toString());
        writer.flush();
        buffer.clear();
    }

    private static class MetricSample {
        final long timestampMs;
        final long elapsedMs;
        final String testName;
        final int iteration;
        final float cpuPercent;
        final float ramMb;
        final float gpuPercent;
        final float fps;
        final int pid;

        MetricSample(long timestampMs, long elapsedMs, String testName, int iteration,
                float cpuPercent, float ramMb, float gpuPercent, float fps, int pid) {
            this.timestampMs = timestampMs;
            this.elapsedMs = elapsedMs;
            this.testName = testName;
            this.iteration = iteration;
            this.cpuPercent = cpuPercent;
            this.ramMb = ramMb;
            this.gpuPercent = gpuPercent;
            this.fps = fps;
            this.pid = pid;
        }
    }
}

File: app/src/main/java/com/jossy/android/mobilebenchmarkappkotlin/metrics/SystemMetricsCollector.kt
package com.jossy.android.mobilebenchmarkappkotlin.metrics

import android.content.Context
import android.os.Debug
import android.os.Handler
import android.os.Looper
import android.os.Process
import android.util.Log
import android.view.Choreographer
import kotlinx.coroutines.*
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.RandomAccessFile
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * Kolektor metryk systemowych - zbiera CPU, RAM, GPU, FPS bezpośrednio w aplikacji.
 * Zapisuje dane do CSV w czasie rzeczywistym z synchronizacją testów.
 */
class SystemMetricsCollector(
    context: Context,
    private val outputDir: File,
    private val sessionId: String
) {
    companion object {
        private const val TAG = "SystemMetricsCollector"
        private const val SAMPLING_INTERVAL_MS = 100L
        private const val BUFFER_SIZE = 50
    }

    private val appContext = context.applicationContext
    private val outputFile = File(outputDir, "system_metrics_$sessionId.csv")
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val mainHandler = Handler(Looper.getMainLooper())

    // Stan
    private val running = AtomicBoolean(false)
    @Volatile private var currentTestName = "IDLE"
    @Volatile private var currentIteration = 0
    private var startTimeMs = 0L

    // Metryki CPU
    private var prevCpuTime = 0L
    private var prevAppCpuTime = 0L

    // Metryki FPS
    private val frameCount = AtomicInteger(0)
    private val lastFpsCalcTime = AtomicLong(0)
    @Volatile private var currentFps = 0f

    // Bufor i writer
    private var writer: BufferedWriter? = null
    private val buffer = mutableListOf<MetricSample>()

    // Choreographer callback
    private var frameCallback: Choreographer.FrameCallback? = null

    data class MetricSample(
        val timestampMs: Long,
        val elapsedMs: Long,
        val testName: String,
        val iteration: Int,
        val cpuPercent: Float,
        val ramMb: Float,
        val gpuPercent: Float,
        val fps: Float,
        val pid: Int
    )

    /**
     * Rozpoczyna zbieranie metryk.
     */
    fun start() {
        if (running.getAndSet(true)) return

        try {
            outputDir.mkdirs()
            writer = BufferedWriter(FileWriter(outputFile), 32 * 1024)
            writer?.write("TimestampMs,ElapsedMs,TestName,Iteration,CPU%,RAM_MB,GPU%,FPS,PID\n")

            startTimeMs = System.currentTimeMillis()
            lastFpsCalcTime.set(startTimeMs)

            startFpsCounter()
            scope.launch { samplingLoop() }

            Log.i(TAG, "Started: ${outputFile.absolutePath}")
        } catch (e: Exception) {
            Log.e(TAG, "Start failed: ${e.message}", e)
            running.set(false)
        }
    }

    /**
     * Zatrzymuje zbieranie metryk.
     */
    fun stop() {
        if (!running.getAndSet(false)) return

        stopFpsCounter()
        scope.cancel()

        try {
            synchronized(buffer) {
                flushBuffer()
            }
            writer?.close()
            writer = null
        } catch (e: Exception) {
            Log.e(TAG, "Stop error: ${e.message}", e)
        }

        Log.i(TAG, "Stopped")
    }

    /**
     * Ustawia aktualny test dla synchronizacji.
     */
    fun setCurrentTest(testName: String, iteration: Int) {
        currentTestName = testName
        currentIteration = iteration
    }

    private suspend fun samplingLoop() {
        while (running.get()) {
            val loopStart = System.currentTimeMillis()

            try {
                val sample = collectSample()
                synchronized(buffer) {
                    buffer.add(sample)
                    if (buffer.size >= BUFFER_SIZE) {
                        flushBuffer()
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Sampling error: ${e.message}")
            }

            val elapsed = System.currentTimeMillis() - loopStart
            val sleepMs = maxOf(0, SAMPLING_INTERVAL_MS - elapsed)
            if (sleepMs > 0) delay(sleepMs)
        }
    }

    private fun collectSample(): MetricSample {
        val now = System.currentTimeMillis()
        return MetricSample(
            timestampMs = now,
            elapsedMs = now - startTimeMs,
            testName = currentTestName,
            iteration = currentIteration,
            cpuPercent = getCpuUsage(),
            ramMb = getMemoryUsageMB(),
            gpuPercent = getGpuUsage(),
            fps = currentFps,
            pid = Process.myPid()
        )
    }

    private fun getCpuUsage(): Float {
        return try {
            val pid = Process.myPid()

            // /proc/[pid]/stat
            RandomAccessFile("/proc/$pid/stat", "r").use { procStat ->
                val statLine = procStat.readLine()
                val parts = statLine.split("\\s+".toRegex())
                if (parts.size < 15) return 0f

                val utime = parts[13].toLong()
                val stime = parts[14].toLong()
                val appCpuTime = utime + stime

                // /proc/stat
                RandomAccessFile("/proc/stat", "r").use { cpuStat ->
                    val cpuLine = cpuStat.readLine()
                    val cpuParts = cpuLine.split("\\s+".toRegex())
                    val cpuTime = (1..7).sumOf { cpuParts.getOrNull(it)?.toLongOrNull() ?: 0L }

                    if (prevCpuTime > 0 && prevAppCpuTime > 0) {
                        val cpuDelta = cpuTime - prevCpuTime
                        val appDelta = appCpuTime - prevAppCpuTime
                        if (cpuDelta > 0) {
                            val cores = Runtime.getRuntime().availableProcessors()
                            val percent = appDelta.toFloat() / cpuDelta * 100f * cores
                            prevCpuTime = cpuTime
                            prevAppCpuTime = appCpuTime
                            return minOf(100f * cores, percent)
                        }
                    }

                    prevCpuTime = cpuTime
                    prevAppCpuTime = appCpuTime
                    0f
                }
            }
        } catch (e: Exception) {
            0f
        }
    }

    private fun getMemoryUsageMB(): Float {
        return try {
            val memInfo = Debug.MemoryInfo()
            Debug.getMemoryInfo(memInfo)
            memInfo.totalPss / 1024f
        } catch (e: Exception) {
            0f
        }
    }

    private fun getGpuUsage(): Float {
        // Qualcomm Adreno
        try {
            File("/sys/class/kgsl/kgsl-3d0/gpu_busy_percentage").readText().trim().toFloatOrNull()?.let { return it }
        } catch (_: Exception) {}

        // Mali
        try {
            File("/sys/class/misc/mali0/device/utilization").readText().trim().toFloatOrNull()?.let { return it }
        } catch (_: Exception) {}

        return -1f
    }

    private fun startFpsCounter() {
        frameCallback = object : Choreographer.FrameCallback {
            override fun doFrame(frameTimeNanos: Long) {
                if (!running.get()) return

                frameCount.incrementAndGet()

                val now = System.currentTimeMillis()
                val lastCalc = lastFpsCalcTime.get()
                if (now - lastCalc >= 1000) {
                    val frames = frameCount.getAndSet(0)
                    val interval = (now - lastCalc) / 1000f
                    currentFps = frames / interval
                    lastFpsCalcTime.set(now)
                }

                Choreographer.getInstance().postFrameCallback(this)
            }
        }

        mainHandler.post {
            Choreographer.getInstance().postFrameCallback(frameCallback!!)
        }
    }

    private fun stopFpsCounter() {
        mainHandler.post {
            frameCallback?.let { Choreographer.getInstance().removeFrameCallback(it) }
        }
    }

    private fun flushBuffer() {
        if (buffer.isEmpty() || writer == null) return

        try {
            val sb = StringBuilder()
            buffer.forEach { s ->
                sb.append(String.format(Locale.US,
                    "%d,%d,%s,%d,%.2f,%.2f,%.2f,%.1f,%d\n",
                    s.timestampMs, s.elapsedMs, s.testName, s.iteration,
                    s.cpuPercent, s.ramMb, s.gpuPercent, s.fps, s.pid))
            }
            writer?.write(sb.toString())
            writer?.flush()
            buffer.clear()
        } catch (e: Exception) {
            Log.e(TAG, "Flush error: ${e.message}")
        }
    }
}
